// @moralisweb3/streams/lib/mapping/logs-processor/LogParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogParser = void 0;
var abi_1 = require("@ethersproject/abi");
var LogParser = /** @class */ (function () {
    function LogParser(abiItems) {
        this.abiInterface = new abi_1.Interface(abiItems);
    }
    LogParser.prototype.read = function (log) {
        // Solidity supports max 3 topics. https://docs.soliditylang.org/en/latest/contracts.html#events
        var topics = [log.topic0, log.topic1, log.topic2, log.topic3].filter(function (t) { return t !== null; });
        var result = this.abiInterface.parseLog({
            data: log.data,
            topics: topics,
        });
        var params = {};
        for (var _i = 0, _a = result.eventFragment.inputs; _i < _a.length; _i++) {
            var input = _a[_i];
            params[input.name] = {
                type: input.type,
                value: result.args[input.name],
            };
        }
        return {
            name: result.name,
            params: params,
        };
    };
    return LogParser;
}());
exports.LogParser = LogParser;
