// @moralisweb3/streams-typings/addresses.ts
/**
 * Stringified UUIDv4.
 * See [RFC 4112](https://tools.ietf.org/html/rfc4122)
 * @pattern [0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-4[0-9A-Fa-f]{3}-[89ABab][0-9A-Fa-f]{3}-[0-9A-Fa-f]{12}
 * @format uuid
 */
export type UUID = string;

export interface AddressesAdd {
  /**
   * The address or a list of addresses to be added to the Stream.
   */
  address: string | string[];
}

export interface AddressesRemove {
  /**
   * The address to be removed from the Stream.
   */
  address: string;
}

export interface AddressesResponse {
  /**
   * Array of project Streams
   */
  result: Addresses[];
  /**
   * Cursor for fetching next page
   */
  cursor?: string;
  /**
   * Total count of streams on the project
   */
  total: number;
}

export interface Addresses {
  /**
   * Address
   */
  address: string;
  /**
   * Unique id of object
   */
  id: UUID;
}

export interface DeleteAddressResponse {
  /**
   * The streamId
   */
  streamId: string;
  /**
   * Address
   */
  address: string;
}

export interface AddressResponse {
  /**
   * The streamId
   */
  streamId: string;
  /**
   * Address
   */
  address: string | string[];
}
