// @moralisweb3/sol-api/lib/resolvers/token/getTokenPrice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTokenPrice = void 0;
var api_utils_1 = require("@moralisweb3/api-utils");
var core_1 = require("@moralisweb3/core");
var sol_utils_1 = require("@moralisweb3/sol-utils");
var SolNetworkResolver_1 = require("../SolNetworkResolver");
exports.getTokenPrice = (0, api_utils_1.createEndpointFactory)(function (core) {
    return (0, api_utils_1.createEndpoint)({
        name: 'getTokenPrice',
        urlParams: ['network', 'address'],
        getUrl: function (params) {
            // TODO: here should be: const network = SolNetworkResolver.resolve(params.network, core);
            // but it's not working with Endpoints.getDescriptors(). After changes described in Endpoints
            // please replace this line.
            var network = params.network ? params.network : SolNetworkResolver_1.SolNetworkResolver.resolve(undefined, core);
            return "/token/".concat(network, "/").concat(params.address, "/price");
        },
        apiToResult: function (data) {
            return {
                nativePrice: {
                    value: sol_utils_1.SolNative.create(data.nativePrice.value, 'solana'),
                    decimals: data.nativePrice.decimals,
                    name: data.nativePrice.name,
                    symbol: data.nativePrice.symbol,
                },
                usdPrice: data.usdPrice,
                exchangeAddress: sol_utils_1.SolAddress.create(data.exchangeAddress),
                exchangeName: data.exchangeName,
            };
        },
        resultToJson: function (data) {
            return {
                nativePrice: {
                    value: data.nativePrice.value.toJSON(),
                    decimals: data.nativePrice.decimals,
                    name: data.nativePrice.name,
                    symbol: data.nativePrice.symbol,
                },
                usdPrice: data.usdPrice,
                exchangeAddress: data.exchangeAddress.toJSON(),
                exchangeName: data.exchangeName,
            };
        },
        parseParams: function (params) {
            var network = SolNetworkResolver_1.SolNetworkResolver.resolve(params.network, core);
            if (network !== 'mainnet') {
                throw new core_1.MoralisApiError({
                    message: "Incorrct value for 'network', getTokenPrice is only available on mainnet",
                    code: core_1.ApiErrorCode.INVALID_PARAMS,
                });
            }
            return {
                network: network,
                address: sol_utils_1.SolAddress.create(params.address).address,
            };
        },
    });
});
