// @parse/push-adapter/lib/APNS.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.APNS = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _nodeApn = require('@parse/node-apn');

var _nodeApn2 = _interopRequireDefault(_nodeApn);

var _parse = require('parse');

var _parse2 = _interopRequireDefault(_parse);

var _npmlog = require('npmlog');

var _npmlog2 = _interopRequireDefault(_npmlog);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var LOG_PREFIX = 'parse-server-push-adapter APNS';

var APNS = exports.APNS = function () {

  /**
   * Create a new provider for the APN service.
   * @constructor
   * @param {Object|Array} args An argument or a list of arguments to config APNS provider
   * @param {Object} args.token {Object} Configuration for Provider Authentication Tokens. (Defaults to: null i.e. fallback to Certificates)
   * @param {Buffer|String} args.token.key The filename of the provider token key (as supplied by Apple) to load from disk, or a Buffer/String containing the key data.
   * @param {String} args.token.keyId The ID of the key issued by Apple
   * @param {String} args.token.teamId ID of the team associated with the provider token key
   * @param {Buffer|String} args.cert The filename of the connection certificate to load from disk, or a Buffer/String containing the certificate data.
   * @param {Buffer|String} args.key {Buffer|String} The filename of the connection key to load from disk, or a Buffer/String containing the key data.
   * @param {Buffer|String} args.pfx path for private key, certificate and CA certs in PFX or PKCS12 format, or a Buffer containing the PFX data. If supplied will always be used instead of certificate and key above.
   * @param {String} args.passphrase The passphrase for the provider key, if required
   * @param {Boolean} args.production Specifies which environment to connect to: Production (if true) or Sandbox
   * @param {String} args.topic Specififies an App-Id for this Provider
   * @param {String} args.bundleId DEPRECATED: Specifies an App-ID for this Provider
   * @param {Number} args.connectionRetryLimit  The maximum number of connection failures that will be tolerated before apn.Provider will "give up". (Defaults to: 3)
   */
  function APNS(args) {
    _classCallCheck(this, APNS);

    // Define class members
    this.providers = [];

    // Since for ios, there maybe multiple cert/key pairs, typePushConfig can be an array.
    var apnsArgsList = [];
    if (Array.isArray(args)) {
      apnsArgsList = apnsArgsList.concat(args);
    } else if ((typeof args === 'undefined' ? 'undefined' : _typeof(args)) === 'object') {
      apnsArgsList.push(args);
    } else {
      throw new _parse2.default.Error(_parse2.default.Error.PUSH_MISCONFIGURED, 'APNS Configuration is invalid');
    }

    // Create Provider from each arg-object
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = apnsArgsList[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var apnsArgs = _step.value;


        // rewrite bundleId to topic for backward-compatibility
        if (apnsArgs.bundleId) {
          _npmlog2.default.warn(LOG_PREFIX, 'bundleId is deprecated, use topic instead');
          apnsArgs.topic = apnsArgs.bundleId;
        }

        var provider = APNS._createProvider(apnsArgs);
        this.providers.push(provider);
      }

      // Sort the providers based on priority ascending, high pri first
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    this.providers.sort(function (s1, s2) {
      return s1.priority - s2.priority;
    });

    // Set index-property of providers
    for (var index = 0; index < this.providers.length; index++) {
      this.providers[index].index = index;
    }
  }

  /**
   * Send apns request.
   *
   * @param {Object} data The data we need to send, the format is the same with api request body
   * @param {Array} allDevices An array of devices
   * @returns {Object} A promise which is resolved immediately
   */


  _createClass(APNS, [{
    key: 'send',
    value: function send(data, allDevices) {
      var coreData = data.data;
      var expirationTime = data['expiration_time'] || coreData['expiration_time'];
      var collapseId = data['collapse_id'] || coreData['collapse_id'];
      var pushType = data['push_type'] || coreData['push_type'];
      var priority = data['priority'] || coreData['priority'];
      var allPromises = [];

      var devicesPerAppIdentifier = {};

      // Start by clustering the devices per appIdentifier
      allDevices.forEach(function (device) {
        var appIdentifier = device.appIdentifier;
        devicesPerAppIdentifier[appIdentifier] = devicesPerAppIdentifier[appIdentifier] || [];
        devicesPerAppIdentifier[appIdentifier].push(device);
      });

      for (var key in devicesPerAppIdentifier) {
        var devices = devicesPerAppIdentifier[key];
        var appIdentifier = devices[0].appIdentifier;
        var providers = this._chooseProviders(appIdentifier);

        // No Providers found
        if (!providers || providers.length === 0) {
          var errorPromises = devices.map(function (device) {
            return APNS._createErrorPromise(device.deviceToken, 'No Provider found');
          });
          allPromises = allPromises.concat(errorPromises);
          continue;
        }

        var headers = { expirationTime: expirationTime, topic: appIdentifier, collapseId: collapseId, pushType: pushType, priority: priority };
        var notification = APNS._generateNotification(coreData, headers);
        var deviceIds = devices.map(function (device) {
          return device.deviceToken;
        });
        var promise = this.sendThroughProvider(notification, deviceIds, providers);
        allPromises.push(promise.then(this._handlePromise.bind(this)));
      }

      return Promise.all(allPromises).then(function (results) {
        // flatten all
        return [].concat.apply([], results);
      });
    }
  }, {
    key: 'sendThroughProvider',
    value: function sendThroughProvider(notification, devices, providers) {
      var _this = this;

      return providers[0].send(notification, devices).then(function (response) {
        if (response.failed && response.failed.length > 0 && providers && providers.length > 1) {
          var _devices = response.failed.map(function (failure) {
            return failure.device;
          });
          // Reset the failures as we'll try next connection
          response.failed = [];
          return _this.sendThroughProvider(notification, _devices, providers.slice(1, providers.length)).then(function (retryResponse) {
            response.failed = response.failed.concat(retryResponse.failed);
            response.sent = response.sent.concat(retryResponse.sent);
            return response;
          });
        } else {
          return response;
        }
      });
    }
  }, {
    key: '_chooseProviders',


    /**
     * Choose appropriate providers based on device appIdentifier.
     *
     * @param {String} appIdentifier appIdentifier for required provider
     * @returns {Array} Returns Array with appropriate providers
     */
    value: function _chooseProviders(appIdentifier) {
      // If the device we need to send to does not have appIdentifier, any provider could be a qualified provider
      /*if (!appIdentifier || appIdentifier === '') {
          return this.providers.map((provider) => provider.index);
      }*/

      // Otherwise we try to match the appIdentifier with topic on provider
      var qualifiedProviders = this.providers.filter(function (provider) {
        return appIdentifier === provider.topic;
      });

      if (qualifiedProviders.length > 0) {
        return qualifiedProviders;
      }

      // If qualifiedProviders empty, add all providers without topic
      return this.providers.filter(function (provider) {
        return !provider.topic || provider.topic === '';
      });
    }
  }, {
    key: '_handlePromise',
    value: function _handlePromise(response) {
      var promises = [];
      response.sent.forEach(function (token) {
        _npmlog2.default.verbose(LOG_PREFIX, 'APNS transmitted to %s', token.device);
        promises.push(APNS._createSuccesfullPromise(token.device));
      });
      response.failed.forEach(function (failure) {
        promises.push(APNS._handlePushFailure(failure));
      });
      return Promise.all(promises);
    }
  }], [{
    key: '_validateAPNArgs',
    value: function _validateAPNArgs(apnsArgs) {
      if (apnsArgs.topic) {
        return true;
      }
      return !(apnsArgs.cert || apnsArgs.key || apnsArgs.pfx);
    }

    /**
     * Creates an Provider base on apnsArgs.
     */

  }, {
    key: '_createProvider',
    value: function _createProvider(apnsArgs) {
      // if using certificate, then topic must be defined
      if (!APNS._validateAPNArgs(apnsArgs)) {
        throw new _parse2.default.Error(_parse2.default.Error.PUSH_MISCONFIGURED, 'topic is mssing for %j', apnsArgs);
      }

      var provider = new _nodeApn2.default.Provider(apnsArgs);

      // Sets the topic on this provider
      provider.topic = apnsArgs.topic;

      // Set the priority of the providers, prod cert has higher priority
      if (apnsArgs.production) {
        provider.priority = 0;
      } else {
        provider.priority = 1;
      }

      return provider;
    }

    /**
     * Generate the apns Notification from the data we get from api request.
     * @param {Object} coreData The data field under api request body
     * @param {Object} headers The header properties for the notification (topic, expirationTime, collapseId, pushType, priority)
     * @returns {Object} A apns Notification
     */

  }, {
    key: '_generateNotification',
    value: function _generateNotification(coreData, headers) {
      var notification = new _nodeApn2.default.Notification();
      var payload = {};
      for (var key in coreData) {
        switch (key) {
          case 'aps':
            notification.aps = coreData.aps;
            break;
          case 'alert':
            notification.setAlert(coreData.alert);
            break;
          case 'title':
            notification.setTitle(coreData.title);
            break;
          case 'badge':
            notification.setBadge(coreData.badge);
            break;
          case 'sound':
            notification.setSound(coreData.sound);
            break;
          case 'content-available':
            var isAvailable = coreData['content-available'] === 1;
            notification.setContentAvailable(isAvailable);
            break;
          case 'mutable-content':
            var isMutable = coreData['mutable-content'] === 1;
            notification.setMutableContent(isMutable);
            break;
          case 'targetContentIdentifier':
            notification.setTargetContentIdentifier(coreData.targetContentIdentifier);
            break;
          case 'interruptionLevel':
            notification.setInterruptionLevel(coreData.interruptionLevel);
            break;
          case 'category':
            notification.setCategory(coreData.category);
            break;
          case 'threadId':
            notification.setThreadId(coreData.threadId);
            break;
          default:
            payload[key] = coreData[key];
            break;
        }
      }

      notification.payload = payload;

      notification.topic = headers.topic;
      notification.expiry = Math.round(headers.expirationTime / 1000);
      notification.collapseId = headers.collapseId;
      // set alert as default push type. If push type is not set notifications are not delivered to devices running iOS 13, watchOS 6 and later.
      notification.pushType = 'alert';
      if (headers.pushType) {
        notification.pushType = headers.pushType;
      }
      if (headers.priority) {
        // if headers priority is not set 'node-apn' defaults it to 5 which is min. required value for background pushes to launch the app in background.
        notification.priority = headers.priority;
      }
      return notification;
    }
  }, {
    key: '_handlePushFailure',
    value: function _handlePushFailure(failure) {
      if (failure.error) {
        _npmlog2.default.error(LOG_PREFIX, 'APNS error transmitting to device %s with error %s', failure.device, failure.error);
        return APNS._createErrorPromise(failure.device, failure.error);
      } else if (failure.status && failure.response && failure.response.reason) {
        _npmlog2.default.error(LOG_PREFIX, 'APNS error transmitting to device %s with status %s and reason %s', failure.device, failure.status, failure.response.reason);
        return APNS._createErrorPromise(failure.device, failure.response.reason);
      } else {
        _npmlog2.default.error(LOG_PREFIX, 'APNS error transmitting to device with unkown error');
        return APNS._createErrorPromise(failure.device, 'Unkown status');
      }
    }

    /**
     * Creates an errorPromise for return.
     *
     * @param {String} token Device-Token
     * @param {String} errorMessage ErrrorMessage as string
     */

  }, {
    key: '_createErrorPromise',
    value: function _createErrorPromise(token, errorMessage) {
      return Promise.resolve({
        transmitted: false,
        device: {
          deviceToken: token,
          deviceType: 'ios'
        },
        response: { error: errorMessage }
      });
    }

    /**
     * Creates an successfulPromise for return.
     *
     * @param {String} token Device-Token
     */

  }, {
    key: '_createSuccesfullPromise',
    value: function _createSuccesfullPromise(token) {
      return Promise.resolve({
        transmitted: true,
        device: {
          deviceToken: token,
          deviceType: 'ios'
        }
      });
    }
  }]);

  return APNS;
}();

exports.default = APNS;
