// @moralisweb3/common-streams-utils/lib/operations/evmStreams/updateStreamEvmOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateStreamEvmOperation = void 0;
var common_evm_utils_1 = require("@moralisweb3/common-evm-utils");
var dataTypes_1 = require("../../dataTypes");
exports.updateStreamEvmOperation = {
    method: 'POST',
    name: 'updateStreamEvm',
    id: 'UpdateStream',
    groupName: 'evmStreams',
    urlPathParamNames: ['id'],
    urlPathPattern: '/streams/evm/{id}',
    bodyParamNames: [
        'webhookUrl',
        'description',
        'tag',
        'topic0',
        'allAddresses',
        'includeNativeTxs',
        'includeContractLogs',
        'includeInternalTxs',
        'chains',
        'abi',
        'advancedOptions',
    ],
    bodyType: 'properties',
    getRequestUrlParams: getRequestUrlParams,
    getRequestBody: getRequestBody,
    deserializeResponse: deserializeResponse,
    serializeRequest: serializeRequest,
    deserializeRequest: deserializeRequest,
};
// Methods
function getRequestUrlParams(request) {
    return {
        id: request.id,
    };
}
function getRequestBody(request, core) {
    return {
        webhookUrl: request.webhookUrl,
        description: request.description,
        tag: request.tag,
        topic0: request.topic0,
        allAddresses: request.allAddresses,
        includeNativeTxs: request.includeNativeTxs,
        includeContractLogs: request.includeContractLogs,
        includeInternalTxs: request.includeInternalTxs,
        chainIds: request.chains.map(function (chain) { return common_evm_utils_1.EvmChain.create(chain, core).apiHex; }),
        abi: request.abi,
        advancedOptions: request.advancedOptions,
    };
}
function deserializeResponse(jsonResponse, request, core) {
    return dataTypes_1.EvmStream.create(jsonResponse, core);
}
function serializeRequest(request, core) {
    return {
        id: request.id,
        webhookUrl: request.webhookUrl,
        description: request.description,
        tag: request.tag,
        topic0: request.topic0,
        allAddresses: request.allAddresses,
        includeNativeTxs: request.includeNativeTxs,
        includeContractLogs: request.includeContractLogs,
        includeInternalTxs: request.includeInternalTxs,
        chainIds: request.chains.map(function (chain) { return common_evm_utils_1.EvmChain.create(chain, core).apiHex; }),
        abi: request.abi,
        advancedOptions: request.advancedOptions,
    };
}
function deserializeRequest(jsonRequest, core) {
    return {
        id: jsonRequest.id,
        webhookUrl: jsonRequest.webhookUrl,
        description: jsonRequest.description,
        tag: jsonRequest.tag,
        topic0: jsonRequest.topic0,
        allAddresses: jsonRequest.allAddresses,
        includeNativeTxs: jsonRequest.includeNativeTxs,
        includeContractLogs: jsonRequest.includeContractLogs,
        includeInternalTxs: jsonRequest.includeInternalTxs,
        chains: jsonRequest.chainIds.map(function (chainId) { return common_evm_utils_1.EvmChain.create(chainId, core); }),
        abi: jsonRequest.abi,
        advancedOptions: jsonRequest.advancedOptions,
    };
}
